Set TEST_HOST and BUNDLE_LOADER for app unit tests

ollieatkinson opened the pull request:
set `TEST_HOST` and `BUNDLE_LOADER` for Unit tests bundle for an app target so that the default generated test target runs

Resolves https://github.com/tuist/tuist/issues/204

### Short description 📝

The default generated project did not set these two build configurations which meant that the app unit test target was not able to resolve the symbols stores inside of the application.

### Solution 📦

By setting `TEST_HOST` and `BUNDLE_LOADER` ld is able to link the symbols and run the unit tests.

The implementation will look at the dependencies for the test target and the find the app to set as the test host. Currently this selects the first app, but we can expand on this in the future as and when we better understand peoples requirements.

### Test Plan 👩‍💻👨‍💻

- [x] Enable all unit tests for applications inside of fixtures
- [x] `bundle exec rake features`
- [x] `swift test`

pepibumur commented on Sources/TuistKit/Generator/ConfigGenerator.swift:
```diff
@@ -165,6 +170,17 @@ final class ConfigGenerator: ConfigGenerating {
         }
 
+        if target.product.isTest {
+            let appDependencies = graph.targetDependencies(path: sourceRootPath, name: target.name).filter { targetNode in
```
Nitpick: You can do 

```swift
let appDependency = find(where: { $0.targetNode.target.product == .app }
```

ollieatkinson replied:
Where is the `find` function? I don't think it's defined.

kwridan replied:
`first(where:` 🙂

kwridan commented:
Nice to this being finally resolved! 👍 

Sadly UI tests targets won't work with this :/ 

<img width="1072" alt="Screen Shot 2019-03-23 at 9 29 39 AM" src="https://user-images.githubusercontent.com/11914919/54864387-68a5e080-4d4e-11e9-94b5-ca8f827cefe6.png">

```swift
Target(name: "AppUITests",
                                 platform: .iOS,
                                 product: .uiTests,
                                 bundleId: "io.tuist.AppUITests",
                                 infoPlist: "Tests.plist",
                                 sources: "UITests/**",
                                 dependencies: [
                                     .target(name: "App"),
                                 ])
```

ollieatkinson commented:
Thanks @kwridan  - I've fixed the issue here: https://github.com/tuist/tuist/pull/300/commits/a57f2e63b6022b85a28a5d914ad55f78592e864a

I set `TEST_TARGET_NAME` and ensured that the test target identity was being set at the top level PBXProject. I also added the ui tests to the `ios_app_with_tests` fixture.

kwridan commented on Sources/TuistKit/Generator/ConfigGenerator.swift:
```diff
@@ -165,6 +170,21 @@ final class ConfigGenerator: ConfigGenerating {
+            }
+
+            if let app = appDependencies.first {
+                settings["TEST_TARGET_NAME"] = "\(app.target.name)"
```
looks like Xcode only populates `TEST_TARGET_NAME` for ui tests only - it doesn't seem to mind it being there, guess this helps simplify `generateTestTargetIdentity`

ollieatkinson replied:
I will leave it, keeps it simple.

kwridan commented on Sources/TuistKit/Generator/ProjectGenerator.swift:
```diff
@@ -166,6 +171,34 @@ final class ProjectGenerator: ProjectGenerating {
+                continue
+            }
+
+            pbxProject.setTargetAttributes(["TestTargetID": target], target: testTarget)
```
This will replace any existing attributes - I guess atm target attributes aren't being added elsewhere but will be hard to spot later on.

```swift
var attributes = pbxProject.targetAttributes[testTarget] ?? [:]
attributes["TestTargetID"] = target
pbxProject.setTargetAttributes(attributes, target: testTarget)
```

worth seeing if we can add a test for target attributes - will take a look

kwridan commented on Tests/TuistKitTests/Generator/ConfigGeneratorTests.swift:
```diff
@@ -79,6 +79,23 @@ final class ConfigGeneratorTests: XCTestCase {
+        assert(config: debugConfig, contains: testHostSettings)
+        assert(config: releaseConfig, contains: testHostSettings)
+    }
+
```
worth adding another case for `uiTests` verifying `TEST_TARGET_NAME`

kwridan commented on features/generate.feature:
```diff
@@ -7,6 +7,7 @@ Feature: Generate a new project using Tuist
     Then tuist generates the project
     Then I should be able to build the scheme App
     Then I should be able to test the scheme AppTests
+    Then I should be able to test the scheme AppUITests
```
👍 nice
